apps/navstar_overlay/src/navstar_overlay_lashup_kv_listener.erl

```diff
@@ -114,68 +116,121 @@ terminate(_Reason, _State, _Data) ->
-    lager:debug("Done applying configuration: ~p", [DeltaConfig]),
-    NewConfig = update_config(DeltaConfig, OldConfig),
-    Data1 = Data0#data{config = NewConfig},
-    {next_state, batching, Data1};
-configuring(info, _EventContent, _Data) ->
+    StateData2 = StateData1#data{kill_timer = Timer},
+    EventContent = #{key => OverlayKey, value => Value},
+    {next_state, configuring, StateData2, {next_event, internal, EventContent}}.
+
+configuring(internal, Config, _StateData) ->
+    lager:debug("Applying configuration: ~p", [Config]),
+    navstar_overlay_configure:start_link(Config),
+    keep_state_and_data;
+configuring(info, {navstar_overlay_configure, applied_config, AppliedConfig}, StateData0) ->
+    lager:debug("Done applying configuration: ~p", [AppliedConfig]),
+    StateData1 = maybe_update_state(AppliedConfig, StateData0),
+    next_state_transition(StateData1);
+configuring(info, _EventContent, _StateData) ->
     {keep_state_and_data, postpone}.
             
-batching(info, EventContent, _Data) ->
-    {keep_state_and_data, {timeout, ?LISTEN_TIMEOUT, {do_configure, EventContent}}};
-batching(timeout, {do_configure, EventContent}, Data) ->
-    {next_state, configuring, Data, {next_event, internal, EventContent}}.
+batching(info, LashupEvent, StateData0) ->
+    StateData1 = handle_event(LashupEvent, StateData0),
+    {keep_state, StateData1, {timeout, ?BATCH_TIMEOUT, do_configure}};
+batching(timeout, do_configure, StateData0 = #data{config = Config}) ->
+    {StateData1, Actions} = next_state_and_actions(Config, StateData0),
+    {next_state, configuring, StateData1, Actions};
+batching(timeout, do_reapply, StateData) ->
+    {next_state , reapplying, StateData, {next_event, internal, reapply}}.
+
+reapplying(internal, reapply, StateData0 = #data{overlaykeys = OverlayKeys}) ->
+    Config = fetch_lashup_config(OverlayKeys),
+    {StateData1, Actions} = next_state_and_actions(Config, StateData0),
+    {next_state, configuring, StateData1, Actions}.
 
 %% private API
 
--spec(delta_config(NewConfig :: config2(), OldConfig :: config()) -> config2()).
-delta_config(NewConfig = #{key := OverlayKey, value := NewValue}, OldConfig) ->
-    case maps:get(OverlayKey, OldConfig, []) of
-        [] ->
-            NewConfig;
-        OldValue ->
-            DeltaValue = ordsets:to_list(ordsets:subtract(ordsets:from_list(NewValue), OldValue)),
-            #{key => OverlayKey, value => DeltaValue}
-    end.
-
--spec(update_config(config2(), OldConfig :: config()) -> config()).
-update_config(#{key := OverlayKey, value := DeltaValue}, OldConfig) ->
-    NewValue = case maps:get(OverlayKey, OldConfig, []) of
-                   [] ->
-                      ordsets:from_list(DeltaValue);
-                   OldValue ->
-                      ordsets:union(ordsets:from_list(DeltaValue), OldValue)
-               end,
-    maps:put(OverlayKey, NewValue, OldConfig).
+-spec(handle_event(LashupEvent :: tuple(), #data{}) -> #data{}).
```

Comment:
Don't use handle_event, it's a special function name when gen_statem is in a different callback mode.